Code changes to fix SeLion Grid issues

sundaramrajendran opened the pull request:
- Fixed the issue which will allow to read SeLionConfig.json from selionHome directory
- Reduced the probability of node registering automatically to the Grid when SeLionRemoteProxy remove node from Grid for forceshutdown

mach6 commented on server/src/main/java/com/paypal/selion/utils/ConfigParser.java:
```diff
@@ -48,6 +48,10 @@
      */
     public static ConfigParser parse() {
+        LOGGER.entering();
+        if (parser == null) {
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Singleton is not thread safe](https://www.codacy.com/app/SeLion/paypal-SeLion/file/1516698859/issues/source?bid=2928425&fileBranchId=3008049#l52)

mach6 replied:
@sundaramrajendran Will you please elaborate on the issue this change is fixing? The code change you made here is a revert of a previous change you made. See https://github.com/sundaramrajendran/SeLion/commit/b8064e0686225b38982510fe1c7f0c72748dcb7c#diff-f0886e3373d41f8f4e2ca67ec83de2e4

sundaramrajendran replied:
@mach6 Yeah I have reverted the previous commit. Issue is SeLion not able to read the SeLionConfig.json file from selionHome directory specified from java arguments because of previous change. Since we are creating static instance for ConfigParser, always it will load the SeLionConfig.json from jar file, not the specified file via setConfigFile(String file). Here ConfigParser getting initialized before we are setting config file. So we need lazy initialization to make it work,

mach6 replied:
@sundaramrajendran Yes, I see the bug now. Thanks!  This said, I think `ConfigParser` needs a few more changes.
1. We should not call `readConfigFileContents` in any constructor.
2. We should wait to call `readConfigFileContents` on the first invocation of `parse`
3. `setConfigFile` should return the `ConfigParser` instance to make it more like a builder. E.g. - `ConfigParser.setConfigFile(file).parse().getInt(key)`
4. Subsequent calls to `setConfigFile` should not change the static value of `configFile` once `parse` has also been called.

sundaramrajendran commented:
@mach6 I have addressed all the review comments except adding new constructor. ConfigParser we are using singleton pattern, So I didn't find any importance of adding constructor with configFile as parameter.

mach6 commented:
@sundaramrajendran sorry on the constructor comment.. I noticed my error and actually edited/removed that suggestion in my note.

mach6 commented:
@sundaramrajendran How about?

mach6 commented:
@sundaramrajendran What about these two suggestions? 

> 1. `setConfigFile` should return the `ConfigParser` instance to make it more like a builder. E.g. - `ConfigParser.setConfigFile(file).parse().getInt(key)`
> 2. Subsequent calls to `setConfigFile` should not change the static value of `configFile` once parse has also been called.

mach6 commented on server/src/main/java/com/paypal/selion/utils/ConfigParser.java:
```diff
@@ -58,9 +66,13 @@ public static ConfigParser parse() {
+    public static ConfigParser setConfigFile(String file) {
         LOGGER.entering(file);
-        configFile = file;
+        if (configuration == null) {
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Singleton is not thread safe](https://www.codacy.com/app/SeLion/paypal-SeLion/file/1536614930/issues/source?bid=2928425&fileBranchId=3008049#l71)

mach6 replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Singleton is not thread safe](https://www.codacy.com/app/SeLion/paypal-SeLion/file/1571905575/issues/source?bid=2928425&fileBranchId=3008049#l71)

mach6 replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Singleton is not thread safe](https://www.codacy.com/app/SeLion/paypal-SeLion/file/1595242861/issues/source?bid=2928425&fileBranchId=3008049#l71)

mach6 commented on server/src/test/java/com/paypal/selion/utils/ConfigParserTest.java:
```diff
@@ -15,26 +15,17 @@
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.Test;
 
-@PrepareForTest(ConfigParser.class)
```
Doesn't removing this mock object in this test class create a race condition for the rest of the test classes
which may rely on the legit config file?  `{selionHome}/config/SeLionConfig.json` vs `/config/DummySeLionConfig.json`

sundaramrajendran replied:
It should, But I didn't face the issue. Let me check once again and update the code if required.

sundaramrajendran commented:
I have fixed the ConfigParserTest with PowerMockito.

sundaramrajendran committed:
Code changes to fix SeLion Grid issues

 - Fixed the node hang issue by avoided SeLionRemoteProxy removing the node from registry when Grid force shutdown the node
 - Fixed the issue which will allow to read SeLionConfig.json from selionHome directory
 - Fixed the failed unit test for ConfigParser
